mgmt: hawkbit: autohandler and shell optimize

maass-hamburg opened the pull request:
also use the workqueue for the run of hawkBit, when requested via the shell. This way the run via the shell will use the same thread as the autohandler run. We also can eliminate redundant code. 

This has mainly the benefit of not having to extend the stack size of the shell threads. Explicitly when having multiple shell instances, then the stack of every instance has to be bigger. 

This is needed for https://github.com/zephyrproject-rtos/zephyr/pull/70787 as it moves the buffers from global to the stack.

maass-hamburg committed:
mgmt: hawkbit: use K_WORK_DELAYABLE_DEFINE

staticly init hawkbit_work_handle

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

ycsin commented on subsys/mgmt/hawkbit/shell.c:
```diff
@@ -18,49 +18,7 @@ static void cmd_run(const struct shell *sh, size_t argc, char **argv)
-		shell_fprintf(sh, SHELL_ERROR, "Invalid response\n");
-		break;
-	}
+	hawkbit_autohandler_once();
```
This is not equal to what we had before - `shell_fprintf` would 'reply' to the shell instance where the command came from, `LOG_*` would be sending to whatever log backends configured with the target verbosity.

So if only LOG UART backend is enabled and command is coming from shell_telnet, the user will not be able to see the messages.

ycsin commented on subsys/mgmt/hawkbit/shell.c:
```diff
@@ -18,13 +18,14 @@ static void cmd_run(const struct shell *sh, size_t argc, char **argv)
 
 	shell_fprintf(sh, SHELL_INFO, "Starting hawkBit run...\n");
-	switch (hawkbit_probe()) {
+	hawkbit_autohandler(false);
```
Are you going to see the same message twice? One from the log and the other one from the shell?

maass-hamburg replied:
yes, if you have the log output activated for that shell. But if you have other log backends activated, like syslog it should also get the logs, independent of the reason why hawkbit was executed.

ycsin replied:
Is there a better way to refactor this mechanism so that it only prints the result to the shell instance that triggered it?

maass-hamburg replied:
IMO it is good, that the autohandler, sending these messages to all log backends, independent of the reason why it was executed (shell or scheduled). It's more like a feature. 
I don't think it is easily possible to omit the logs from the autohandler just for the shell in which it was started, while still sending the log to all other log backends, including the other shells, when they are enabled as log backend.

maass-hamburg commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -92,7 +92,9 @@ static union {
-static struct k_work_delayable hawkbit_work_handle;
+static void autohandler(struct k_work *work);
+
+static K_WORK_DELAYABLE_DEFINE(hawkbit_work_handle, autohandler);
```
so that it isn't initialised in `hawkbit_autohandler()` and `hawkbit_autohandler()` can be executed more that one without reinitialising again. For maybe converting hawkBit to be instance-based, which will need lot of changes, initialising it statically is not the problem, it would be this being a global variable and that's not changing with this PR. Also making hawkbit instance-based would not be the optimal solution, as hawkBit already supports having multiple software modules as part of a distribution, where one module is the OS and another can be the for settings. That would also be much easier to implement.

maass-hamburg committed:
mgmt: hawkbit: also use workqueue for shell run

Also use a workqueue, when execution of
hawkBit is requested via shell.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
mgmt: hawkbit: add option for autohandler

add option for autohandler to only run once.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
mgmt: hawkbit: add hawkbit_autohandler_wait

Add hawkbit_autohandler_wait() to be able
to wait for the autohandler to finish.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
mgmt: hawkbit: log start of autohandler from shell

If the run of the autohandler is started from shell,
it will be logged.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
mgmt: hawkbit: delay autohandler

Be able to delay the next run of
the hawkbit autohandler.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

kartben commented on include/zephyr/mgmt/hawkbit.h:
```diff
@@ -88,8 +89,52 @@ int hawkbit_init(void);
+ * @param if_bigger If true, the delay will be set only if the new delay is bigger than the current
+ * one.
+ *
+ * @return 0 if @a if_bigger was true and the current delay was bigger than the new one.
+ * @return otherwise, a value from k_work_reschedule().
```
use `@retval`

kartben commented on include/zephyr/mgmt/hawkbit.h:
```diff
@@ -88,8 +89,52 @@ int hawkbit_init(void);
+ * @return HAWKBIT_PERMISSION_ERROR fail to get the permission to access the hawkBit server.
+ * @return HAWKBIT_METADATA_ERROR fail to parse or to encode the metadata.
+ * @return HAWKBIT_DOWNLOAD_ERROR fail while downloading the update package.
+ * @return HAWKBIT_OK if success.
```
make this one first in the list?

maass-hamburg commented on include/zephyr/mgmt/hawkbit.h:
```diff
@@ -48,8 +49,30 @@ int hawkbit_init(void);
+ * @param timeout Waiting period for the desired set of events or one of the
+ *                special values K_NO_WAIT and K_FOREVER.
+ *
+ * @return HAWKBIT_NO_RESPONSE if matching events were not received within the specified time
+ * @return HAWKBIT_NETWORKING_ERROR fail to connect to the hawkBit server.
+ * @return HAWKBIT_UNCONFIRMED_IMAGE image is unconfirmed.
+ * @return HAWKBIT_PERMISSION_ERROR fail to get the permission to access the hawkBit server.
+ * @return HAWKBIT_METADATA_ERROR fail to parse or to encode the metadata.
+ * @return HAWKBIT_DOWNLOAD_ERROR fail while downloading the update package.
+ * @return HAWKBIT_OK if success.
+ * @return HAWKBIT_UPDATE_INSTALLED has an update available.
+ * @return HAWKBIT_NO_UPDATE no update available.
+ * @return HAWKBIT_CANCEL_UPDATE update was cancelled.
```
`retval` would be the better one here. https://www.doxygen.nl/manual/commands.html#cmdretval
`retval` and `return` are not the same.

kartben replied:
What @maass-hamburg said :)
See the difference between the two below: 

![image](https://github.com/zephyrproject-rtos/zephyr/assets/128251/73c61232-cae6-4e5b-bf55-f53e0519de3e)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/128251/1dac3c54-88d1-4721-9f1a-abce323ddc76)

ycsin commented on subsys/mgmt/hawkbit/hawkbit.c:
```diff
@@ -1482,12 +1490,27 @@ static void autohandler(struct k_work *work)
 	}
 }
+
+enum hawkbit_response hawkbit_autohandler_wait(uint32_t events, k_timeout_t timeout)
```
Do you expect anyone would use anything other than `__UINT32_MAX__` for `events`?

maass-hamburg replied:
I can think of some people only using `BIT(HAWKBIT_NO_UPDATE)`

maass-hamburg committed:
doc: releases: 4.0: change of hawkbit autohandler

Mention change of hawkbit autohandler and shell in the
migration guide and the release notes.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg commented:
Rebased to resolve conflicts.

aescolar commented on doc/releases/release-notes-3.7.rst:
```diff
@@ -300,6 +300,24 @@ Libraries / Subsystems
+      The autohandler always runs in the system workqueue.
+
+    * With :c:func:`hawkbit_autohandler_wait` you can wait for the autohandler to finish. This
+      function will block until the autohandler is finished.The ``event`` parameter can be passed
+      ``UINT32_MAX`` to wait for the autohandler to finish one run, or passed :c:macro:`BIT`
```
Nit
```suggestion
      function will block until the autohandler is finished. The ``event`` parameter can be set to
      ``UINT32_MAX`` to wait for the autohandler to finish one run, or :c:macro:`BIT`
```
Anyhow, the whole "This  function will block until the autohandler is finished" seems a bit redundant.

aescolar reviewed (commented):
ok'ish for the release notes and migration guide.
Just a minor suggestion for improvement.
Though, if it were me, I would probably be less verbose and just list which new APIs have been added and rely on the link to their doxygen generated docs, instead of reexplaining it in the release notes.

aescolar commented on doc/releases/release-notes-3.7.rst:
```diff
@@ -300,6 +300,24 @@ Libraries / Subsystems
+    * With :c:func:`hawkbit_autohandler_wait` you can wait for the autohandler to finish. This
+      function will block until the autohandler is finished.The ``event`` parameter can be passed
+      ``UINT32_MAX`` to wait for the autohandler to finish one run, or passed :c:macro:`BIT`
+      together with a value from c:enum:`hawkbit_response` to wait for a specific event.
```
```suggestion
      together with a value from :c:enum:`hawkbit_response` to wait for a specific event.
```

maass-hamburg commented:
ping @ycsin

maass-hamburg committed:
mgmt: hawkbit: get poll interval

Add function to get the poll interval.
This is needed to seperate the autohandler
from the main hawkbit code.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
mgmt: hawkbit: seperate autohandler

seperate the autohandler from the main
hawkbit source. This way the autohandler can
be disabled if it is not needed.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
mgmt: hawkbit: move HAWKBIT_JSON_URL

move HAWKBIT_JSON_URL out of the header.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
mgmt: hawkbit: seperate header files

seperate the hawkbit header files, to make
it clearer.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

ycsin reviewed (commented):
I saw that you made some changes to the autohandler stuff, will take a look at those later

Meanwhile here are some quick comments.

The commits should try to explain why you did something in a summary if possible, msg like `"seperate hawkbit.h into different headers."` describes what you did (which we can see from the commit itself), so it didn't provide extra values to people wondering why is this change required

maass-hamburg committed:
doc: releases: 4.0: change of hawkbit header

Mention change of the hawkbit header files.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

ycsin reviewed (dismissed):
I like the fact that we do not need a big stack size for the shell in this PR, but this comes at the expense of multiple additional public functions to meddle with how the hawkBit schedule itself, I'd much rather steer clear of potential racing issues that may arise in the future.

IMHO, while this potentially provides more features to the users, my impression with the built-in autohandler & the shell is that they are there for users to get started with something simple. What I've done in the past is that I'll create my own hawkbit task to do exactly what I need.

My concern with these new features is that they look very 'application' to me (the hawkBit sample basically duplicated the autohandler), I'd prefer hawkBit to stay as a platform/framework/library and let the users decide what's best for them.

From that perspective, these features that we add upstream could restrict the way user uses the hawkBit subsystem - I don't think we will ever satisfy everyone and therefore my expectation is that this trend will likely continue - adding more public functions, each does a very specific thing, attempting to tie up the loose ends, which we will have to maintain.

I understand that this PR is significant as it unblocks #70787 (I didn't really look at that in detail yet, but SMF sounds interesting to me). But if the autohandler or the shell is ever in your way moving forward, my suggestion is that we should just guard them with a Kconfig now and deprecate them in the future. We can demonstrate the features of hawkBit by improving the hawkBit sample.

Would deprecating them unblock #70787?

maass-hamburg commented on include/zephyr/mgmt/hawkbit/autohandler.h:
```diff
@@ -0,0 +1,73 @@
+ *
+ * @param auto_reschedule If true, the handler will reschedule itself
+ */
+void hawkbit_autohandler(bool auto_reschedule);
```
the hawkbit api in zephyr is currently not marked as stable (https://docs.zephyrproject.org/latest/develop/api/overview.html), which means we don't have to follow the normal deprecation process.

ycsin replied:
Interesting, looks like a lot of stuff there could use an update @nashif

maass-hamburg commented on include/zephyr/mgmt/hawkbit/config.h:
```diff
@@ -0,0 +1,176 @@
+	return hawkbit_get_config().tls_tag;
+}
+
+/**
+ * @brief Get the hawkBit action id.
+ *
+ * @return Action id.
+ */
+int32_t hawkbit_get_action_id(void);
+
+/**
+ * @brief Get the hawkBit poll interval.
+ *
+ * @return Poll interval.
+ */
+uint32_t hawkbit_get_poll_interval(void);
```
`hawkbit_get_action_id` is not a new function, only `hawkbit_get_poll_interval` is, so I would like to leave it like it is.

maass-hamburg replied:
this requested change is dismissed, reason above.

maass-hamburg commented on include/zephyr/mgmt/hawkbit/config.h:
```diff
@@ -0,0 +1,176 @@
+	sec_tag_t tls_tag;
+};
+
+/**
+ * @brief Set the hawkBit server configuration settings.
+ *
+ * @param config Configuration settings to set.
+ * @retval 0 on success.
+ * @retval -EAGAIN if probe is currently running.
+ */
+int hawkbit_set_config(struct hawkbit_runtime_config *config);
+
+/**
+ * @brief Get the hawkBit server configuration settings.
+ *
+ * @return Configuration settings.
+ */
+struct hawkbit_runtime_config hawkbit_get_config(void);
```
It is now what is is.

maass-hamburg commented:
rebased to resolve conflicts

aescolar commented:
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

maass-hamburg commented:
I like to get forward with this PR. 
First of all as a Maintainer I want to say that:
- the autohandler and the hawkBit shell will stay. 
- useful additions to both of them are welcome.

Both the shell and the autohandler are a great at making hawkBit easy to use and accessible. 
Not everybody want to write their own autohandler and therefore use the autohandler, which will just work and you don't have to do much yourself. And fore the ones, who don't want to use this autohandler, they can, if this PR gets merged, deactivate the included autohandler. 

That being said, I hope we can get this PR approved and merged soon, so we can finally be able to get to https://github.com/zephyrproject-rtos/zephyr/pull/70787

maass-hamburg commented:
rebased to resolve conflicts

maass-hamburg commented:
@ycsin are you coming back for a review?
In the dev-review meeting in the last week there where no objections against my changes.

maass-hamburg commented:
rebased due to unrelated fails in the ci

ycsin commented on subsys/mgmt/hawkbit/Kconfig:
```diff
@@ -33,9 +33,17 @@ config HAWKBIT_POLL_INTERVAL
 	  This time interval is zero and 43200 minutes(30 days). This will be overridden
 	  by the value configured in the settings of the hawkBit server.
 
+config HAWKBIT_AUTOHANDLER
+	bool "hawkBit autohandler"
+	select EVENTS
+	default y
+	help
+	  Activate autohandler to handle the update process automatically.
+
 config HAWKBIT_SHELL
 	bool "hawkBit shell utilities"
 	depends on SHELL
+	depends on HAWKBIT_AUTOHANDLER
```
> And fore the ones, who don't want to use this autohandler, they can, if this PR gets merged, deactivate the included autohandler.

Does this dependency means that for user who doesn't want to use the bundled autohandler, they lose access to shell for updating the configurations as well?

maass-hamburg replied:
Well yes at least when they don't activate the Kconfig for the autohandler. They could still choose to include the autohandler and don't use it theirself.

ycsin reviewed (commented):
> @ycsin are you coming back for a review?

Had another look but there seems to be nothing new to review, I'd say my previous review is still up-to-date

> First of all as a Maintainer I want to say that:
> 
> * the autohandler and the hawkBit shell will stay.
> * useful additions to both of them are welcome.
> 
> Both the shell and the autohandler are a great at making hawkBit easy to use and accessible. Not everybody want to write their own autohandler and therefore use the autohandler, which will just work and you don't have to do much yourself. And fore the ones, who don't want to use this autohandler, they can, if this PR gets merged, deactivate the included autohandler.

I agree that the bundled autohandler in this PR might be useful, I have no qualms if this PR updated the one in the samples to make it more portable, which IMHO serves pretty much the same purpose to show the users how to build one, instead of forcing an implementation on them.

I've had my custom implementation of the autohandler in the past, had I chose to upstream and developed it around my use case, folks that came after me probably have to either use it my way or go away, I think that would make hawkBit less inviting

maass-hamburg commented:
> > @ycsin are you coming back for a review?
> 
> Had another look but there seems to be nothing new to review, I'd say my previous review is still up-to-date
> 
> > First of all as a Maintainer I want to say that:
> > 
> > * the autohandler and the hawkBit shell will stay.
> > * useful additions to both of them are welcome.
> > 
> > Both the shell and the autohandler are a great at making hawkBit easy to use and accessible. Not everybody want to write their own autohandler and therefore use the autohandler, which will just work and you don't have to do much yourself. And fore the ones, who don't want to use this autohandler, they can, if this PR gets merged, deactivate the included autohandler.
> 
> I agree that the bundled autohandler in this PR might be useful, I have no qualms if this PR updated the one in the samples to make it more portable, which IMHO serves pretty much the same purpose to show the users how to build one, instead of forcing an implementation on them.
> 
> I've had my custom implementation of the autohandler in the past, had I chose to upstream and developed it around my use case, folks that came after me probably have to either use it my way or go away, I think that would make hawkBit less inviting

@ycsin As I said before the autohandler and the shell are going to stay and if there are good reasons, like here, they are also going to be extended. Requested changes in the other way are hereby dismissed. You can follow the [escalation process](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-technical-escalation) if you are not okay with that, otherwise I expect a review approving this or requesting changes that are not on this fundamental level.

maass-hamburg commented:
@jukkar @rlubos can you maybe take a look?

jukkar reviewed (approved):
I am not very familiar with hawkbit so my review is just looking the overall code structure. Anyway, the changes look reasonable.

maass-hamburg commented:
@kartben there are still requested changes from you, you might want to approve?
